auto adjust valid gpu config

psychocrypt opened the pull request:
The current `config.txt`is not running on all devices. It could be hard for non advanced users to find a first valid configuration.

This pull request adds the possibility that the miner adjust the device setting to find a valid run able configuration. The user can start to tune the settings based on the auto adjust settings.
The auto adjust function is **not** searching for the best setting, only a valid configuration based on the gpu type, operating system and currently available memory is searched.

## Changes
- allow `threads`, `blocks`, `bfactor` and `bsleep` to be auto adjusted by the miner
  - default on windows:
    - `bsleep=25`
    - `bfactor=6`
  - default on linux
    - `bsleep=0`
    - `bfactor`6`
  - default `blocks` 3 * SMX (multi processor count)
  - default `threads <=64` auto selected depends on the architecture and the available gpu memory
- change default `config.txt` to use auto adjusted gpu value
- add class `autoAdjust`
- ~~print used device configuration with and memory usage with `debug_level : 3` (important information to tune the miner to the maximum hash rate)~~
- ~~compile CUDA as C++11 (missed to enable it in #10)~~

## Tests
- [x] runtime 2 different device in one system (both auto adjust)
- [x] runtime 4 devices full and partly auto adjusted

fireice-uk commented:
I feel like evicting this code to a separate function and extending it to print the whole config table. You wouldn't believe how many people are confused by it.

psychocrypt commented:
Yes we can do, is the jansson reader you are using able to write jansson formated output to the terminal?
Keep in mind that this pull request also print the memory footprint. This is a very helpful information to get the maximum hash rate out of the miner.


There is one issue with writing the full config out, each thread is handling it's own gpu and if we not add a global barrier to sync all worker threads we don't know the point where all gpus are adjusted.
On a system with 8 gpus it take around 30 seconds to initialize all gpus.

Could you help me with a jansson terminal printer to show the full config.
It is also possible to overwrite the user config file if a `-1` is included and at the adjusted values.

psychocrypt commented:
Other suggestion: we add a verbose level where the current L3 and the adjust information is shown. If the user configered all he ca change the lvl to L3 and not get spammend anymore with this information.

fireice-uk commented:
My view for the function would be to:
- Detect null in config table [default setting] and remove config count altogether
- If that is the case we branch into system scan mode
- We print a roughly correct config
- Exit allowing the user to copy and paste it himself

While automagic json writing can be done, it will screw up newlines and look horrible,

psychocrypt commented:
What do you mean with null, the auto adjust value `-1`?


How would we handle the printed memory footprints?
A new option in the config file?

fireice-uk commented:
`
- /* 
-  * Number of GPUs that you have in your system. Each GPU will get its own CPU thread.
-  */
- "gpu_thread_num" : 1,
- 
/*
 * GPU configuration. You should play around with threads and blocks as the fastest settings will vary.
 * index         - GPU index number usually starts from 0.
 * threads       - Number of GPU threads (nothing to do with CPU threads).
 * blocks        - Number of GPU blocks (nothing to do with CPU threads).
 * bfactor       - Enables running the Cryptonight kernel in smaller pieces. 
 *                 Increase if you want to reduce GPU lag. Recommended setting on GUI systems - 8
 * bsleep        - Insert a delay of X microseconds between kernel launches. 
 *                 Increase if you want to reduce GPU lag. Recommended setting on GUI systems - 100
 * affine_to_cpu - This will affine the thread to a CPU. This can make a GPU miner play along nicer with a CPU miner.
 */
"gpu_threads_conf" : null,
`

Remove the lines starting with minus. Of course this will require a pretty much complete rewrite of this pr.

psychocrypt commented:
I force pushed a new version with the changes:
  - removed the memory foot print from this pull request (I will add it as feature for the next version after the relase)
  - I added the null directly to the values (`blocks`, `threads` ...) that a advanced user can modify the values directly without the need to start the miner. If we mark the group with null the documentation above looks wrong because the user can't find the value `threads` within the config file.

## Test
- [ ] runtime test

fireice-uk commented:
@psychocrypt
Problems in rough order of significance:
- It uses the number of configs specified by the user rather than the one needed
- I want to get rid of printf and use printer instead
- Press any key is only relevant on Windows - it is better to use win_exit()

Would you like me to do this one? I can do it a little bit later on tonight after I finish a proper fetcher for Monero wallet.

psychocrypt commented:
> It uses the number of configs specified by the user rather than the one needed

I would prefer to implement a auto suggestion for the number of devices in a separate pull request after this one. But I can do it also within this one.

> I want to get rid of printf and use printer instead

The printer always prints the timestamp before, this is not copy past able. Therefore I switched to printf. I would prefere std::cout but formating is a pain. Should I extent the printer to allow printing without the timestep?

> Press any key is only relevant on Windows - it is better to use win_exit()

This is also needed on linux if someone start the miner by klick on the binary and select eun in terminal. The terminal closed and the user can't see the asjusted config.

I can add changes tomorrow or you later

fireice-uk commented:
Ok. I will do the broad strokes today (modify jconf and printer), and you can finish off the more gpu-based code and test tomorrow.

fireice-uk commented:
Also my bad, printer function is already there `print_str`

psychocrypt commented:
I uploaded new changes
  - use printer instead of `printf`
  - handle if jansson group `gpu_threads_conf` is null
  - add cuda fucntion to get cuda device count
  - change calculation of `device_blocks` for <sm_30

I format the adjuster output that there is no invalid line break on a linux cmd (those would result in errors after copy&past).

- [ ] test windows
- [ ] test linux

fireice-uk commented:
Merging for testing.
